java/ql/src/semmle/code/SMAP.qll

```diff
@@ -0,0 +1,55 @@
+/**
+ * Provides classes and predicates for working with SMAP files (see JSR-045).
+ */
+
+import java
+
+/**
+ * Holds if there exists a mapping between an SMAP input file and line
+ * and a corresponding SMAP output file and line range.
+ */
+private predicate smap(File inputFile, int inLine, File outputFile, int outLineStart, int outLineEnd) {
+  exists(
+    string defaultStratum, int inputFileNum, int inStart, int inCount, int outStart, int outIncr,
+    int n
+  |
+    smap_header(outputFile, _, defaultStratum) and
+    smap_files(outputFile, defaultStratum, inputFileNum, _, inputFile) and
+    smap_lines(outputFile, defaultStratum, inputFileNum, inStart, inCount, outStart, outIncr) and
+    inLine in [inStart .. inStart + inCount - 1] and
+    outLineStart = outStart + n * outIncr and
+    outLineEnd = (n + 1) * outIncr - 1 + outStart and
+    n = inLine - inStart
+  )
+}
+
+/**
+ * Holds if there exists a mapping between an SMAP input file and line
+ * and a corresponding SMAP output file and line.
+ */
+private predicate smap(File inputFile, int inLine, File outputFile, int outLine) {
+  exists(int outLineStart, int outLineEnd |
+    smap(inputFile, inLine, outputFile, outLineStart, outLineEnd) and
+    outLine in [outLineStart .. outLineEnd]
+  )
+}
+
+/**
+ * Holds if an SMAP input location (with path, line and column information)
+ * has a corresponding SMAP output location (with path and line information).
+ *
+ * For example, an SMAP input location may be a location within a JSP file,
+ * which may have a corresponding SMAP output location in generated Java code.
+ */
+predicate hasSmapLocationInfo(
+  string inputPath, int isl, int isc, int iel, int iec, string outputPath, int osl, int oel
+) {
+  exists(File inputFile, File outputFile |
+    inputPath = inputFile.getAbsolutePath() and
+    outputPath = outputFile.getAbsolutePath() and
+    smap(inputFile, isl, outputFile, osl) and
+    smap(inputFile, iel - 1, outputFile, oel) and
```

Comment:
This is a cartesian product of all line mappings with itself for a particular file-pair. This has the potential to blow up for medium-sized mapped files. To prevent this we should add some manual magic in the form of `locations_default(_, _, osl, _, oel, _)`. We should also inspect how the combination of this addition and the two `smap` conjuncts join-orders, and we should verify that this manual magic doesn't remove any tuples from its use in `hasLocationInfo`.  We can possibly also magic in the file in the `locations_default` conjunct, but I don't think that's necessarily needed.

Reply:
Including the file in the magic conjunct cuts the size down further, i.e. using `locations_default(_, outputFile, osl, _, oel, _)`, so on second thought we probably should include it.  The end result is a classic 3-way join, so we'll need to make certain that the magic conjunct isn't chosen last (as that will produce the cartesian product as intermediate tuples).

Reply:
Thanks for spotting this and pushing a fix.